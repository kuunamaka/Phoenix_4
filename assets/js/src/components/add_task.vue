<!-- adding a new task -->
<template>
  <table>
    <h2>Adding Task</h2>
    <div>
      <div>
        Name
        <input v-model="name">
      </div>
      <div>
        Assigned by:
        <select v-model="assignee_id">
          <option disabled>Please select one</option>
          <option v-for="task in tasks" :key="task.id">
            {{ task.assignee_id }}
          </option>
        </select>
      </div>
      <div>
        Already finished?
        <input type="checkbox" id="is_finished" v-model="is_finished">
      </div>
    </div>
    <button @click="createTask">Submit</button>
    <a href="/tasks" class="button">Back</a>
  </table>
</template>

<script>
import axios from 'axios'
export default {
  name: 'add_task',
  data() {
    return {
      name: '',
      assignee_id: null,
      is_finished: false,
      tasks: null
    }
  },
  methods: {
    async createTask() {
      try {
        await axios.post('/api/tasks', {
          task: {
            name: this.name,
            is_finished: this.is_finished,
            assignee_id: this.assignee_id
          }
        })
        window.location.href = '/tasks';
      }
      catch (e) {
      }
    }
  },
  async mounted() {
    try {
      const resp = await axios.get('/api/tasks')
      this.tasks = resp.data.tasks
    }
    catch (e) {
    }
  }
}
</script>

