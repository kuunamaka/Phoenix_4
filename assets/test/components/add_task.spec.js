import { shallowMount } from '@vue/test-utils'
import _ from 'lodash'
import add_task from  '@/src/components/add_task.vue'

const axios = require("axios");
const MockAdapter = require("axios-mock-adapter");
const mock = new  MockAdapter(axios);

global.window = Object.create(window);
const pathname = '/tasks/new';

Object.defineProperty(window, 'location', {
  value: {
    pathname,
    href: `http://localhost:4001${ pathname }`
  }
});

describe('add_task.vue', () => {
  const wrapper = shallowMount(add_task);
  const vm = wrapper.vm
  
  it('renders entire component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('createTask', () => {
    it('redirect to /tasks', async () => {  
      mock.onPost("/api/tasks", {task: {name: 'Test', is_finished: false, assignee_id: 1}}).reply(200);
      await wrapper.setData({name: 'Test', is_finished: false, assignee_id: 1})
        
      await vm.createTask()

      expect(window.location.href).toEqual('/tasks');
    });
  });

  it('show validate error messages', async () => {
    mock.onPost("/api/tasks", {task: {name: '', is_finished: false, assignee_id: 1}}).reply(404);
    await wrapper.setData({name: '', is_finished: false, assignee_id: 1});

    const createButton = wrapper.find('button');
    await createButton.trigger('click');

    expect(wrapper.find('p').element).toMatchSnapshot();
  });
});
