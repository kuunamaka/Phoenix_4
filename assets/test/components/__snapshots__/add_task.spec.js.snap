// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add_task.vue is a Vue instance 1`] = `
<table>
  <h2>
    Adding Task
  </h2>
   
  <div>
    <!---->
     
    <div>
      
      Name
      
      <input />
    </div>
     
    <div>
      
      Assigned by:
      
      <select>
        <option
          disabled="disabled"
        >
          Please select one
        </option>
         
      </select>
    </div>
     
    <div>
      
      Already finished?
      
      <input
        id="is_finished"
        type="checkbox"
      />
    </div>
  </div>
   
  <button>
    Submit
  </button>
   
  <a
    class="button"
    href="/tasks"
  >
    Back
  </a>
</table>
`;

exports[`add_task.vue show validate error messages 1`] = `
<p>
  
      Please correct the following error(s):
      
  <ul>
    <li
      class="error"
    >
       Name required
    </li>
  </ul>
</p>
`;
